protocols/bsbport/bsbport_mqtt.c

```diff
@@ -0,0 +1,252 @@
+      mqtt_construct_publish_packet(topic, buf, len, BSBPORT_MQTT_RETAIN);
+      BSBDEBUG("%s=%s", topic, buf);
+
+      // TMP
+      snprintf_P(topic, TOPIC_LENGTH, publish_topic_format,
+                 LO4(bsbport_msg_buffer.msg[i].src),
+                 bsbport_msg_buffer.msg[i].p.data.p1,
+                 bsbport_msg_buffer.msg[i].p.data.p2,
+                 bsbport_msg_buffer.msg[i].p.data.p3,
+                 bsbport_msg_buffer.msg[i].p.data.p4, "TMP");
+      len =
+        itoa_fixedpoint(((int32_t) bsbport_msg_buffer.msg[i].value * 100) /
+                        64, 2, buf, DATA_LENGTH);
+      mqtt_construct_publish_packet(topic, buf, len, BSBPORT_MQTT_RETAIN);
+      BSBDEBUG("%s=%s", topic, buf);
+
+      // FP1
+      snprintf_P(topic, TOPIC_LENGTH, publish_topic_format,
+                 LO4(bsbport_msg_buffer.msg[i].src),
+                 bsbport_msg_buffer.msg[i].p.data.p1,
+                 bsbport_msg_buffer.msg[i].p.data.p2,
+                 bsbport_msg_buffer.msg[i].p.data.p3,
+                 bsbport_msg_buffer.msg[i].p.data.p4, "FP1");
+      len =
+        itoa_fixedpoint(bsbport_msg_buffer.msg[i].value, 1, buf, DATA_LENGTH);
+      mqtt_construct_publish_packet(topic, buf, len, BSBPORT_MQTT_RETAIN);
+      BSBDEBUG("%s=%s", topic, buf);
+
+      // FP5
+      snprintf_P(topic, TOPIC_LENGTH, publish_topic_format,
+                 LO4(bsbport_msg_buffer.msg[i].src),
+                 bsbport_msg_buffer.msg[i].p.data.p1,
+                 bsbport_msg_buffer.msg[i].p.data.p2,
+                 bsbport_msg_buffer.msg[i].p.data.p3,
+                 bsbport_msg_buffer.msg[i].p.data.p4, "FP5");
+      len =
+        itoa_fixedpoint(bsbport_msg_buffer.msg[i].value * 10 / 2, 1, buf,
+                        DATA_LENGTH);
+      mqtt_construct_publish_packet(topic, buf, len, BSBPORT_MQTT_RETAIN);
+      BSBDEBUG("%s=%s", topic, buf);
+    }
+  }
+}
+
+void
+bsbport_publish_cb(char const *topic, uint16_t topic_length,
+                   const void *payload, uint16_t payload_length)
+{
+  BSBDEBUG("MQTT Publish: %s", topic);
+  if (topic_length < 20)
+    return;
+  uint8_t p1;
+  uint8_t p2;
+  uint8_t p3;
+  uint8_t p4;
+  uint8_t dest = 0;
+  uint8_t ret;
+  char type[4];
+  int16_t fp_val = 0;
+  uint16_t raw_val = 0;
+
+  if (strncmp_P(topic + 4, PSTR("set"), 3) == 0)
+  {
+    char *strvalue = malloc(payload_length + 1);
+    ret =
+      sscanf_P(topic, PSTR(BSBPORT_SUBSCRIBE_SET_FORMAT), &dest, &p1, &p2,
+               &p3, &p4, type);
+    memcpy(strvalue, payload, payload_length);
+    strvalue[payload_length] = 0;
+    sscanf_P(strvalue, PSTR("%i"), &raw_val);
+    next_int16_fp(strvalue, &fp_val, 1);
+
+    BSBDEBUG("MQTT set ARGS:%d %02x %02x %02x %02x %02x %3s %s %d %u", ret,
+             p1, p2, p3, p4, dest, type, strvalue, fp_val, raw_val);
+
+    if (ret == 6)
+    {
+      uint8_t data[3];
+      uint8_t datalen = 3;
+      data[0] = 0x01;
+      if (strcmp_P(type, PSTR("RAW")) == 0)
```

Comment:
How about using the first charachter to indentify the type? If ambiguous, test the next character?